README.md
Arbitrary Precision
===================

I wrote this in college. Many of the algorithms are not the best, but it should have some educational value.

This is an Integer class that can handle arbitrarily large integers. It should handle numbers into the millions of digits. Upon this is build a decimal floating point class that should be able to handle computations up to 256 digits of precision (this limitation is due to magic numbers in the library routines). That's a far cry from the 17 you get from the hardware with IEEE-754. Read Description.txt for more information.

Yes, this is also hosted on BitBucket. I forgot the password to that account years ago.

Programs included:
* AltSlowCalc - An RPN calculator with decimal floating-point numbers. It is the "SlowCalc" that is mentioned in the files under GPL3. "SlowCalc" was a college project that had a less efficient decimal floating-point number implementation.
* DB12 - A crappy language interpreter, made from my compiler class project.
* Dice - A tool for generating the actual number of cases for building the probabilities with very large rolls of dice.
* Example - Another crappy language interpreter.
* FENZero4Float - A tool for finding the zeros of a one-dimensional functions. No, it doesn't use Newton's method (I've never successfully written code to do differentiation), so it is MUCH slower, but can find zeros that Newton's method cannot.
* IntCalc - An integer calculator that is only notable in that it implements Dijkstra's Shunting-Yard Algorithm, an algorithm I will never again implement because I understand recursive descent parsers now.


Changes.txt
So I have been trying to make IntCalc faster, specifically in calculating
2^2^20. IntCalc originally took around 14 minutes to compute this and display
it, while GNU bc takes around 14 seconds (on my machine). My first step was to
modify BitField. BitField was originally written to be memory conservative and
manage memory meticulously. I rewrote it so that the bit data was not
contained in it, but contained in another structure that did reference
tracking. BitField became Copy-On-Write, and did some more optimizations:
it no longer reallocated memory when a number required less than was
allocated, and anticipates carries when it creates a copy during addition.
I also changed it from storing 15 bits in a 16 bit short to storing 32 bits in
a 32 bit long. The overhead of type promotion seems to lose over the less
operations that are performed due to the greater storage density. Moving on to
Integer, I got the best speed performance increase by changing toString() and
fromString() to pull and push "fat digits" from the BigInt. In decimal,
toString() pulls out nine digits from the BigInt, then breaks them apart. It
does one long division and nine small divisions, instead of nine long
divisions. At this point, I started improving the functionality of the
calculator, which led me to find a bug in compare(). My next project was to
implement word division, which is apparently slightly faster. Division was
also changed to have defined behavior should quotient and remainder be
aliased, as it had undefined behavior before (undefined as in could explode).
I keep comparing IntCalc to bc, and overall, either it is slower at addition
and subtraction, or the overhead of converting to/from overwhelms any speed
improvement. It IS slower at multiplication. I have no intention, at this
point, of implementing Karatsuba Multiplication as bc does, and O(n^2) can't
beat O(n^1.6). Division, on the other hand, is a win (At least on the level of
dividing 200 digit numbers by 100 digit numbers, twice, as I did 100K times
during a test [The test was actually to ensure the correctness of division,
but I couldn't help but time it]). Both binary and word division are both
faster than bc's decimal division.
After all of these changes, 2^2^20 (that's 2^1048576 not 2^40) now takes about
50 seconds with IntCalc, a dramatic improvement.

2011/01/04

I found a bug in the division code. It's easy to reproduce: 2^1024/2^512 = 0.
A '>' should have been '>='. Decided to add the Fibbonacci numbers as an
addition, and an extra post-fix "operator".
Remember: fib(n+1) / fib(n) * 2 - 1 = sqrt(5) for very large values of n.
Try: 500F*10^50/499F*2-10^50; Mentally place a '.' after the initial 2.

I did some math (!) and found that division and multiplication should
altogether cease to work at around 1 billion digit numbers. This is a
limitation of using 32 bit ints to specify shift values. Such numbers would be
using megabytes of memory, at that point. Addition, subtraction, and
single word multiplication and division (which are used to convert to/from
strings) should still work beyond that, in theory, to the limit of memory.

2011/01/05

After looking at other people's code, I found a better solution for Newton-
Rhapson sqrt() in William Rossi's code. I implemented it instead. What I was
using was optimized for floating-point calculations, not integer.

2011/01/19

I decided to rewrite the loop in BitField::operator*= to be more symmetric with
the one in BitField::operator/=. Hopefully, it will be easier for optimizers
to understand.

2011/01/21

I fixed a crippling division bug, where division would hang on the right
input. Try S(2*10^20); to see the error. I then tried to make it always
generate a leading 0, but that is as bad as generating a trailing 0. I finally
found the winning combination of code to make things work correctly: and it
always generates a leading 0.

2011/01/23

I decided to make Integer a true binary bigint, and make it do lots of funny
stuff and be independant. I moved the functions into the Integer namespace,
and added alot of useful functions.

2011/01/26

So, I didn't even do a test compile on 1/26, then I tried to use the file on
May 5, and had to fix it. This incorporates all of the fixes, mostly random
misuse of member data. There was also a bug in the Integer >> and <<, where a
negative shift wasn't turned into a positive shift in the other direction. The
only thing that isn't included is the functions to turn strings into Integers
and back agian. Most people won't use them. Finally, I added the divide
tempering test from Knuth, instead of running a loop.

2011/05/20

Fixed a stupid bug in toString. Now, it returns an empty string when base is
not in [2,36], instead of trying to do undefined math. Fixed a bug in
divApprox, a wierd off-by-one error that escaped the 100K tests.

2011/05/22

Changed some questionable logic in Bitfield |= and ^=. We'll see if I broke
anything.

2011/07/09

Made what are mostly cosmetic changes in 32-bit code, but should allow the
code to be compiled correctly under LP64 64-bit code semantics.

2011/07/10

I compiled under 64-bit GCC today. Success. With -Wconversion, the only
warnings generated are ignorable: until we have 256-bit computers, I could
use an unsigned char for the ints in question. Thinking about it, the code
should also work under LLP64 semantics (Microsoft), but it looses the only
benefit of 64 bit code: large address space.

2011/07/11

Fixed some instances where -0 could have been generated.

2011/07/13

Changed BitField::operator>>= to do less work, so it should be faster.

2011/12/22

Fixed a silly mistake in Integer::operator--(int).
It was still the pre-decrement operation.

2012/07/16



IntCalc.cpp
/*
Copyright (c) 2010, 2011 Thomas DiModica.
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
1. Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.
3. Neither the name of Thomas DiModica nor the names of other contributors
   may be used to endorse or promote products derived from this software
   without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THOMAS DIMODICA AND CONTRIBUTORS ``AS IS'' AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED.  IN NO EVENT SHALL THOMAS DIMODICA OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
SUCH DAMAGE.
*/

/*
   Integer calculator. Supports the operations of addition, subtraction,
   multiplication, integer division, modulo, exponentiation, and factorial.
   Allow for arbitrarily nested parenthesis, and the use of brackets and
   braces as parenthesis. It cannot, however, differentiate between brackets,
   braces, and parenthesis when reporting an error. Use ';' as a stop, a break
   between calculations. Due to how data is read, spaces cannot be used to
   separate calculations.
*/

#include <iostream>
#include <string>
#include <stack>
#include <cstdlib>
#include "Integer.hpp"

using namespace std;
using BigInt::Integer;

const int Ops = 20;

 /*
   If the last token read was a number or a right parenthesis, then a minus
   should be a binary operator.
   If the last token read was an operator or a there is no last token, then a
   minus is either part of a number or a unary operator.
 */
bool LastNum = false;

 /*
   Have we had a "recoverable" error somewhere?
 */
bool Error = false;

/*
   Turn the ASCII value of the token read into a token number.
   Return the index of the token in "+-*""/%^!()~PCNMGLI|SF", if one of these.
   Return 98 if '@'.
   Return 99 if 0-9a-z.
   Else return -1.
*/
const int TokenTable [256] =
 {
   -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
   -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
   -1,  6, -1, -1, -1,  4, -1, -1,  7,  8,  2,  0, -1, 01, -1,  3,
   99, 99, 99, 99, 99, 99, 99, 99, 99, 99, -1, -1, -1, -1, -1, -1,
   98, -1, -1, 11, -1, -1, 19, 14, -1, 16, -1, -1, 15, 13, 12, -1,
   10, -1, -1, 18, -1, -1, -1, -1, -1, -1, -1,  7, -1,  8,  5, -1,
   -1, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99,
   99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99,  7, 17,  8,  9, -1,
   -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
   -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
   -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
   -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
   -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
   -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
   -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
   -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1
 };

 /*
   Returns the associativity of the operator.
   True if Left-to-Right, false otherwise.
   The tenth operator is unary negatation.
 */
const bool LRAssoc [Ops] =
 {
   /* + */ true,
   /* - */ true,
   /* * */ true,
   /* / */ true,
   /* % */ true,
   /* ^ */ false,
   /* ! */ false,
   /* ( */ true,
   /* ) */ true,
   /* ~ */ true,
   /* P */ true,
   /* C */ true,
   /* N */ true,
   /* M */ true,
   /* G */ true,
   /* L */ true,
   /* I */ true,
   /* | */ true,
   /* S */ true,
   /* F */ false
 };

 /*
   We'll give permute/combine the highest precedence and see how that goes.
   I'm just making crap up for the other "operators" as they should be
   functions.
 */
const int Precedence [Ops] =
 {
   /* + */ 0,
   /* - */ 0,
   /* * */ 1,
   /* / */ 1,
   /* % */ 1,
   /* ^ */ 2,
   /* ! */ 3,
   /* ( */ -1,
   /* ) */ -1,
   /* ~ */ 0,
   /* P */ 4,
   /* C */ 4,
   /* N */ 7,
   /* M */ 7,
   /* G */ 6,
   /* L */ 6,
   /* I */ 6,
   /* | */ 0,
   /* S */ 5,
   /* F */ 8
 };

 /*
   Do we have a Binary or Unary operation?
 */
const bool Binary [Ops] =
 {
   /* + */ true,
   /* - */ true,
   /* * */ true,
   /* / */ true,
   /* % */ true,
   /* ^ */ true,
   /* ! */ false,
   /* ( */ false,
   /* ) */ false,
   /* ~ */ false,
   /* P */ true,
   /* C */ true,
   /* N */ true,
   /* M */ true,
   /* G */ true,
   /* L */ true,
   /* I */ true,
   /* | */ false,
   /* S */ false,
   /* F */ false
 };

 /*
   Parse the input LALR(1).
 */
int Token, NextToken;

 /*
   End of computation character.
 */
const int EOC = ';';

void GetFirstToken (void)
 {
   char temp;
   cin >> temp;
   if (!(!cin))
    {
      Token = temp;
      cin >> temp;
      if (!(!cin)) NextToken = temp;
      else NextToken = -1;
    }
   else
    {
      Token = -1;
      NextToken = -1;
    }
 }

void GetNextToken (void)
 {
   char temp;
   Token = NextToken;
   if (Token == EOC) return;
   if (NextToken != -1)
    {
      cin >> temp;
      if (!(!cin)) NextToken = temp;
      else NextToken = -1;
    }
 }

string GetString (void)
 {
   string result;
   result = (char) Token;
   while ((NextToken != -1) && (TokenTable[NextToken] == 99))
    {
      GetNextToken();
      result += (char) Token;
    }
   return result;
 }

Integer operate (int operation, Integer * A, Integer * B)
 {
   Integer result, remainder;
   switch (operation)
    {
      case  0: return *A + *B;
      case  1: return *A - *B;
      case  2: return *A * *B;
      case  3: return div(*A, *B);
      case  4: return mod(*A, *B);
      case  5: return pow(*A, *B);
      case  6: return fact(*A);
      case  9: return -(*A);
      case 10: return permutation(*A, *B);
      case 11: return combination(*A, *B);
      case 12: return min(*A, *B);
      case 13: return max(*A, *B);
      case 14: return gcd(*A, *B);
      case 15: return lcm(*A, *B);
      case 16: return modInv(*A, *B);
      case 17: return abs(*A);
      case 18: return sqrt(*A);
      case 19: return fib(*A);
    }

   cerr << "Internal Error: " << __LINE__ << ". Aborting Calculator." << endl;
   exit(0);
 }

Integer pop (stack <Integer> & me)
 {
   Integer tos;
   if (me.empty())
    {
      cerr << "Parse error: operator without operands." << endl;
      Error = true;
      return Integer ((BigInt::Unit) 0);
    }
   tos = me.top();
   me.pop();
   return tos;
 }

 /*
   I'm going to try to implement Dijkstra's Shunting Yard Algorithm, with the
   exception that there is not going to be an output queue. Numbers will be
   pushed to a data stack, and operations will be performed on the data stack
   when the operators are popped off of the operator stack. So, it is the
   algorithm and the execution of the result at the same time. It just
   requires an extra stack.
 */
int main (int argc, char ** argv)
 {
   stack <Integer> DataStack;
   stack <int> OperatorStack;

   Integer lhs, rhs, last;
   int ibase = 10, obase = 10;

   if (argc > 1)
    {
      obase = atoi(argv[1]);
      if (argc > 2) ibase = atoi(argv[2]);

      if ((ibase < 2) || (ibase > 36) || (obase < 2) || (obase > 36))
       {
         cerr << argv[0] << " [output_base [input_base]]" << endl;
         cerr << "Input and output bases must be greater than one"
            " and less than thirty-seven." << endl;
       }
    }

   cerr << "Integer Playground by Thomas DiModica." << endl
        << "Copyright 2010, 2011." << endl
        << "This software comes with ABSOLUTELY NO WARRANTY." << endl;

   GetFirstToken();

   while (Token != -1)
    {

      do
       {
          // Parse the current Token
         switch (TokenTable[Token])
          {
            case  0:
                // '|' is absolute value. If we would interpret a '+' as such,
                // turn it into such. If the absolute value begins a number,
                // just read it in as part of the number.
               if (!LastNum)
                {
                   // I know everything that they say about goto, but does
                   // anyone see a better way to do this?
                  if (TokenTable[NextToken] == 99) goto ReadANumberInstead;
                  else Token = '|';
                }

               goto FlowThrough;

            case  1:
                // '~' is unary negate. If we would interpret a '-' as such,
                // turn it into such. If the unary negate begins a number,
                // just read it in as part of the number.
               if (!LastNum)
                {
                  if (TokenTable[NextToken] == 99) goto ReadANumberInstead;
                  else Token = '~';
                }

FlowThrough:    // Flow through to generalized operation.

            case  2: // These all flow through, as operation
            case  3: // is abstracted to encompass all operators.
            case  4:
            case  5:
            case  6:
            case  9:
            case 10:
            case 11:
            case 12:
            case 13:
            case 14:
            case 15:
            case 16:
            case 17:
            case 18:
            case 19:
               while (!OperatorStack.empty())
                {
                   // Do we evaluate the operator on the stack?
                  if (LRAssoc[TokenTable[Token]])
                     if (Precedence[TokenTable[Token]] >
                         Precedence[OperatorStack.top()])
                        break;
                  if (!LRAssoc[TokenTable[Token]])
                     if (Precedence[TokenTable[Token]] >=
                         Precedence[OperatorStack.top()])
                        break;

                   // Evaluate the operator on the stack.
                  if (Binary[OperatorStack.top()])
                   {
                     rhs = pop(DataStack);
                     lhs = pop(DataStack);
                     DataStack.push(operate(OperatorStack.top(), &lhs, &rhs));
                   }
                  else
                   {
                     lhs = pop(DataStack);
                     DataStack.push(operate(OperatorStack.top(), &lhs, NULL));
                   }
                  OperatorStack.pop();
                }

               OperatorStack.push(TokenTable[Token]);
               LastNum = false;
               break;

            case  7:

               if (LastNum)
                {
                  cerr << "Parse error: number after number." << endl;
                  Error = true;
                  break;
                }

               OperatorStack.push(TokenTable[Token]);
               LastNum = false;
               break;

            case  8:
               while (!OperatorStack.empty() && (OperatorStack.top() != 7))
                {
                  if (Binary[OperatorStack.top()])
                   {
                     rhs = pop(DataStack);
                     lhs = pop(DataStack);
                     DataStack.push(operate(OperatorStack.top(), &lhs, &rhs));
                   }
                  else
                   {
                     lhs = pop(DataStack);
                     DataStack.push(operate(OperatorStack.top(), &lhs, NULL));
                   }
                  OperatorStack.pop();
                }

               if (OperatorStack.empty())
                {
                  cerr << "Parse error: unmatched parenthesis." << endl;
                  Error = true;
                  break;
                }

               OperatorStack.pop();

               LastNum = true;
               break;

            case 98:

               if (LastNum)
                {
                  cerr << "Parse error: number after number." << endl;
                  Error = true;
                  break;
                }

               DataStack.push(last);
               LastNum = true;
               break;

            case 99:
ReadANumberInstead:

               if (LastNum)
                {
                  cerr << "Parse error: number after number." << endl;
                  Error = true;
                  break;
                }

               DataStack.push(Integer(GetString(), ibase));
               LastNum = true;
               break;

            case -1:
               cerr << "Bad token in token stream." << endl;
               Error = true;
               break;

            default:
                // We should NEVER reach this code.
               cerr << "Internal error: " << __LINE__
                    << ". Aborting calculator." << endl;
               return 0;
          }

         GetNextToken();

         if (Error)
          {
            while ((Token != -1) && (Token != EOC)) GetNextToken();
            Error = false;
          }
       }
      while ((Token != -1) && (Token != EOC));

       // Finalize the parsing
      while (!OperatorStack.empty())
       {
         if (OperatorStack.top() == 7)
          {
            cerr << "Parse error: unmatched parenthesis." << endl;
            break;
          }

         if (Binary[OperatorStack.top()])
          {
            rhs = pop(DataStack);
            lhs = pop(DataStack);
            DataStack.push(operate(OperatorStack.top(), &lhs, &rhs));
          }
         else
          {
            lhs = pop(DataStack);
            DataStack.push(operate(OperatorStack.top(), &lhs, NULL));
          }
         OperatorStack.pop();
       }

       // Print the result
      if (!DataStack.empty()) lhs = DataStack.top();
      else lhs = Integer((BigInt::Unit) 0);

      cout << lhs.toString(obase) << endl;

       // Clear the stacks, just to be sure. They SHOULD be empty right now.
      while (!DataStack.empty()) DataStack.pop();
      while (!OperatorStack.empty()) OperatorStack.pop();

       // Reset this. It's important.
      LastNum = false;
      last = lhs;

       // More input?
      GetFirstToken();
    }

   return 0;
 }
